Prevent multiple threads exporting data at the same time

wargcm opened the pull request:
Fixes "Open WordPress" crash from https://github.com/wordpress-mobile/WordPress-iOS/pull/19787#issuecomment-1356025562

## Description

When WordPress launches, it tries to export data automatically during the startup sequence. The deep link from Jetpack that opens WordPress also tries to export data. I think this can cause a crash when multiple threads are exporting data at the same time.

![Screen Shot 2022-12-19 at 12 16 27 PM](https://user-images.githubusercontent.com/2454408/208524960-b071ba43-b79a-41f1-a0e9-5b34c5c952db.png)

## Testing

### Setup

To get both apps to kick off the export at the same time:

- Comment out [this code](https://github.com/wordpress-mobile/WordPress-iOS/blob/release/21.4/WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift#L97-L100) and install WordPress
- Login

**OR**

- Install an older WordPress such as 21.2 first from TestFlight (so the data export flag isn't set yet)
- Login
- Install over that version with a newer build with the automatic launch off (Edit scheme... > Run > Launch = "Wait for the executable to be launched")

### Steps

To test:

- Setup WordPress following one of the steps from above ^
- Install Jetpack
- Tap on "Open WordPress" when the popup appears
- Verify no crash happens when WordPress opens
- Attempt multiple times since race conditions are fun like that 🙃

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

wargcm committed:
Prevent multiple threads exporting data at the same time

wargcm commented:
I _think_ this is the cause of the crash which is why this is a draft. But I haven't seen it occur again after these changes.

twstokes commented:
Hey @wargcm! 👋 Thanks for tackling this. I wanted to run an idea by you to see if it makes any sense.

**From what I understand, exports typically run as a result of user interaction:**
- Try to export data on WordPress app launch
- Try to export data and send the user to Jetpack (initiated from Jetpack)
- Try to export data and send the user to Jetpack (switch button pressed in WordPress)
- Try to export the data and then delete the exported data if the user is no longer eligible (e.g. site removed)

And currently if any of these exports run at the same time, we can crash due to lack of thread safety. This PR aims to fix this problem by only allowing the export to run once, and upon completion it executes queued up completion handlers.

**Alternative idea:**

1. Keep exporting exclusive so it only runs once (e.g. an `NSLock` around the export process like you have it in this PR)
2. Instead of queuing up completion handlers, immediately throw an error to the losing caller (e.g an export was in progress)
3. In the case of exporting data on WordPress app launch, check the [`UIApplicationLaunchOptionsURLKey`](https://developer.apple.com/documentation/uikit/uiapplicationlaunchoptionsurlkey) to determine whether it was from the Jetpack app or something else like SpringBoard

The justification for this idea is:
- It may slightly reduce complexity by not storing an array of completion handlers
- We won't have to think about completion handler behaviors running consecutively (what if two try to send to Jetpack?)
- It tries to keep an export process tied to a single user action

Do you see any value in that proposal, or any downsides?

wargcm commented:
@twstokes Yep, I could definitely check what launches the app. I can add that here if it still makes sense after the rest of my comment.

I'm a little concerned about immediately calling the completion blocks. There's potential to call a completion handler before the data is finished exporting resulting in unexpected behavior. Like if the completion block opens Jetpack, it could open it right away without the data being ready.

I thought about just removing the lock and only fixing the app launch issue from Jetpack. But there are a couple places off the top of my head where we could potentially run into two exports running at the same time:

- App launch + user tapping on the banner
- When data get reexported (from removing a site / logging out) + tapping on the banner 

Now how realistically the above scenarios can happen is questionable. I tried to profile (debug, so not super accurate) the `startAndDo` call on my device (iPhone SE 3rd gen) and it was taking about ~180 ms for the call to complete. So for me to try one of the above scenarios, it'd basically be impossible. But I also don't have a very big database and my device is relatively newer so I'm not sure how long it could take for other users or devices.

twstokes commented:
All good thoughts @wargcm!

> I'm a little concerned about immediately calling the completion blocks.

Understood. My initial thinking was that the losing calls / immediate completions would carry over an `exportInProgress` failure and we could react accordingly (retry after delay, show a message, do nothing), but I see how this can be tricky.

> I thought about just removing the lock and only fixing the app launch issue from Jetpack. But there are a couple places off the top of my head where we could potentially run into two exports running at the same time:

💯 

If you'd like to add the app launch check I'll be happy to test it.

staskus commented:
I could reproduce the crash consistently when the WordPress app is killed before opening it from Jetpack and this fix works ✅. 

I added logs when testing and I confirm that the guard case of `lock.try()` is hit, an additional completion handler is appended into an array and later called.

```
2022-12-20 11:30:58:922 Test: Locking
2022-12-20 11:30:58:960 didFinishLaunchingWithOptions state: UIApplicationState(rawValue: 1)
2022-12-20 11:30:58:971 Application launched with URL: wordpressmigration+v1://export-213
<...>
2022-12-20 11:30:58:972 Test: Appending completion handler
2022-12-20 11:30:59:068 <WordPress.WordPressAppDelegate: 0x2811acd00> applicationDidBecomeActive(_:)
2022-12-20 11:30:59:104 🔵 Tracked: migration_content_export_succeeded <>
<...>
2022-12-20 11:30:59:146 Test: Local completion
2022-12-20 11:30:59:148 Test: Array completion
```

```
        guard lock.try() else {
            if let completion {
                DDLogError("Test: Appending completion handler")
                completionHandlers.append(completion)
            }
            return
        }
        
        DDLogError("Test: Locking")

        let wrappedCompletion: ((Result<Void, ContentMigrationCoordinatorError>) -> Void) = { [weak self] result in
            guard let self else {
                return
            }
            self.lock.unlock()
            completion?(result)
            DDLogError("Test: Local completion")
            for completionHandler in self.completionHandlers {
                completionHandler(result)
                DDLogError("Test: Array completion")
            }

            self.completionHandlers = []
        }
```

CC: @wargcm

dvdchr commented:
👋🏼 Thanks for wrangling this, @wargcm! Some thoughts about @twstokes 's points:

> exports typically run as a result of user interaction

This is true 💯 . I suspect that the only part where we run the export automatically _and_ asynchronously is what's causing the race condition:

https://github.com/wordpress-mobile/WordPress-iOS/blob/5d89914a31628dfae6e61c207f344ecb59716e76/WordPress/Classes/System/WordPressAppDelegate.swift#L336-L340

Every other export process is triggered by user interaction. Also, they are executed on the main queue synchronously. So if the process happens to take a few seconds, it will block the main thread for that duration. While it's not great if this happens, this should, _in theory_, prevent the possibility of race conditions with other export triggers except the one in App Delegate.

> 2. Instead of queuing up completion handlers, immediately throw an error to the losing caller (e.g., an export was in progress)

One problem is when the one that runs asynchronously gets called first, and the WordPress export deeplink gets triggered second. We should prioritize the completion block from the deeplink as it will bring the user back to Jetpack. 

In fact, the one in the App Delegate should have the lowest priority — if a user interaction triggers an export, we should always give way / prioritize for that one instead. But I'm not sure if this is technically trivial to implement... so this brings up your idea no. 3:

> 3. In the case of exporting data on WordPress app launch, check the [UIApplicationLaunchOptionsURLKey](https://developer.apple.com/documentation/uikit/uiapplicationlaunchoptionsurlkey) to determine whether it was from the Jetpack app or something else like SpringBoard

This makes sense. In the happy path, when the WP app is launched due to the export deeplink, the export in App Delegate won't be executed — thus giving way to the export process initiated from the deeplink. Additionally, we can check the `host` of the deeplink (i.e., if it's the one specifically used for migration), so we can still allow export-on-launch to run when WP is launched through the other links.

@wargcm:
> I thought about just removing the lock and only fixing the app launch issue from Jetpack.

How would you fix the app launch issue from Jetpack? 🤔

wargcm committed:
Revert "Prevent multiple threads exporting data at the same time"

This reverts commit aeb9a3c2a8d2c6631b74955ecf341bea907acb89.

wargcm committed:
Skip exporting data if launched by the export data route

wargcm commented:
@staskus Thanks for testing! I wasn't 100% sure the crash was due to the race condition so that helps.

@twstokes @dvdchr I reverted my changes with the lock stuff and added a check to the app delegate. If we end up seeing the crash again in the wild, then we can add the lock back. I'm assuming that we won't though.

dvdchr commented:
@wargcm sounds good to me! 👍🏼 ~~Also, I think this PR is quite ready for review. What do you think?~~ Nevermind, I was late by 8 minutes. 😅

twstokes commented:
👀 Reviewing this now. 🙇

twstokes reviewed (approved):
This LGTM @wargcm! 🚀 

Testing steps:
1. Commented out [the lines highlighted](https://github.com/wordpress-mobile/WordPress-iOS/blob/release/21.4/WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift#L97-L100) in the testing steps
2. I also added a check for WordPress to the same area, since it'll be introduced [by this commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/19787/commits/42789939adc085145cd14416742b89845447bf21)
3. Ran and logged into WordPress, then exited it (so an initial export wasn't ran)
4. Ran Jetpack and tapped the button to open WordPress
5. Attached the debugger to WordPress as it opened
6. Confirmed that `launchUrl` and `exportUrl` were the same
7. Didn't observe any crashes

spencertransier commented:
This fix made it into the `21.4.0.1` build.
